### PR TITLE
Add registering and login with OTP through email

### PR DESCRIPTION
**Pull request by josepegea**
Add new auth method, using just email and an OTP link.

Based in the work of @phoet in https://github.com/weg-li/weg-li

Both for registering and for logging in, the user just inputs their email and receives a message with a OTP link.

Users registered through this provider lack `nickname` , `name` and `image` on creation. 
In order to deal with the validations for these properties, they're initialized as follows:

- `nickname` is generated by hashing the email address.

- `name` is filled in with a known marker for "missing name" declared in `User::EMPTY_NAME` (currently `"********"`). It doesn't use the email address as a placeholder in order to avoid exposing it, as the user name is displayed in several pages in the application.

- `image` is set to the Gravatar URL for the email.

In order to urge users registered through email to provide a name that can be used in the app, after log in, a user with such an "empty" name is redirected to the profile edit page and asked to provide one.

As with any other source, if such a user later adds another provider, the `name` and `image` coming from those will take over the existing ones.

Should help deal with the Twitter problem described in #1063

### CODE REVIEW

**Comment by JoschkaSchulz** [comment]
@josepegea it looks a bit hacky :) But I understand the problem. Is it really necessary to add the Actionmailer Cop? Couldn't we just solve it, that way we don't have to worry about it in the future?

@salzig what are you thinking?

**Comment by josepegea** [comment]
> @josepegea it looks a bit hacky :) But I understand the problem. Is it really necessary to add the Actionmailer Cop? Couldn't we just solve it, that way we don't have to worry about it in the future?
> 
> @salzig what are you thinking?

Indeed, there are many "hacky" parts about this. 

Truth be told, I feel the user system is too influenced by Twitter being the first login method supported, and not having `email` as the main, unique, identifier has made this addition quite more complex than it should. It's only ironic that now it's Twitter, too, who pushes us to change 😅

As for the `Metrics/ParameterLists` cop, I have my own personal opinion about the overuse of Rubocop and I would gladly remove this one. Making the parameters list shorter here would imply either passing an opaque hash or having the mailer itself deal with getting the data from Whitelabel, and I feel that both options would make the solution more difficult to follow. I was just assuming that the set of cops was already agreed upon and was honoring it. But if we're all fine with removing it, I'll be happy to do that

File: config/locales/de.yml [hunk 1]
```diff
@@ -172,3 +172,17 @@ de:
       previous: "<"
       next: ">"
       truncate: "..."
+  email_auth:
+    header: "Mit E-Mail anmelden"
+    enter_email: "Bitte geben Sie Ihre E-Mail-Adresse ein"
+    submit: "Autorisieren"
+    email_sent: "Eine E-Mail wurde an %{email} mit Details zum Einloggen gesendet"
+    invalid_email: "Bitte geben Sie eine gültige E-Mail-Adresse ein"
+    subject: "Anmeldung zu %{label}"
```
Comment (salzig) [inline config/locales/de.yml:181]:
```suggestion
    subject: "Anmeldung bei %{label}"
```

File: lib/omni_auth/strategies/email.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+      end
+
+      def callback_phase
+        token = request.params['token']
+        fail!(:authenticity_error) if token.blank?
+
+        begin
+          decoded_token = EmailAuthToken.decode(token)
+        rescue StandardError
+          fail!(:authenticity_error, $ERROR_INFO)
```
Comment (salzig) [inline lib/omni_auth/strategies/email.rb:25]:
```suggestion
        rescue StandardError => error
          fail!(:authenticity_error, error)
```
and we can skip the `require 'English'`

File: spec/controllers/sessions_controller_spec.rb [hunk 3]
```diff
@@ -29,10 +29,46 @@
...
+
+  context 'for email auth' do
+    before do
+      request.env['omniauth.auth'] = EMAIL_AUTH_HASH
+      request.host = 'hamburg.onruby.de'
+    end
+
+    context 'POST :email_login' do
+      it 'sends the email and redirect to index', :aggregate_failures do
+        expect { post :email_login, params: { email: 'user@somewhere.org' } }
```
Comment (salzig) [inline spec/controllers/sessions_controller_spec.rb:47]:
```suggestion
        expect { post :email_login, params: { email: 'user@example.org' } }
```
Somewhere.org is registered by "New Radio and Performing Arts, Inc.", that isn't a domain we should use. :) 
http://example.org

File: spec/models/user_spec.rb [hunk 4]
```diff
@@ -57,6 +57,12 @@
       expect(user.email).to eql('phoetmail@googlemail.com')
     end
 
+    it 'sets email and empty name for users created by email' do
+      user = User.create_from_hash!(EMAIL_AUTH_HASH)
+      expect(user.email).to eql('user@somewhere.org')
```
Comment (salzig) [inline spec/models/user_spec.rb:62]:
```suggestion
      expect(user.email).to eql('user@example.org')
```

File: spec/support/factories/authorizations.rb [hunk 5]
```diff
@@ -53,6 +53,14 @@
   },
 }.freeze
 
+EMAIL_AUTH_HASH = {
+  'provider' => 'email',
+  'uid' => 'f8a846d3199e68bb5a7d9ba9dab54ddc9c28e382b5c4490b396214808d26aeb9',
+  'info' => {
+    'email' => 'user@somewhere.org',
```
Comment (salzig) [inline spec/support/factories/authorizations.rb:60]:
```suggestion
    'email' => 'user@example.org',
```

File: spec/models/user_spec.rb [hunk 6]
```diff
@@ -85,6 +91,12 @@
         expect(it.url).to eql('http://blog.nofail.de')
       end
     end
+
+    it 'updates only the email from email-auth-hash', :aggregate_failures do
+      expect { user.update_from_auth!(EMAIL_AUTH_HASH) }
+        .not_to(change { user.attributes.slice(:name, :github, :image, :location, :description, :url) })
+      expect(user.email).to eq('user@somewhere.org')
```
Comment (salzig) [inline spec/models/user_spec.rb:98]:
```suggestion
      expect(user.email).to eq('user@example.org')
```

**Comment by salzig** [comment]
That's my initial 2cents about it. Sorry for the delay.

File: config/locales/de.yml (hunk 1 above)
Comment (josepegea) [reply config/locales/de.yml:181]: I was trusting ChatGPT too much! 😁

**Comment by josepegea** [comment]
> That's my initial 2cents about it. Sorry for the delay.

Thanks for the review! Updated

File: config/initializers/omniauth.rb [hunk 7]
```diff
@@ -1,3 +1,5 @@
+require_relative File.join(Rails.root, 'lib/omni_auth/strategies/email')
```
Comment (phoet) [inline config/initializers/omniauth.rb:1]: no biggie, but the require_relative should handle the directory resolution

File: config/locales/de.yml [hunk 8]
```diff
@@ -172,3 +172,17 @@ de:
       previous: "<"
       next: ">"
       truncate: "..."
+  email_auth:
+    header: "Mit E-Mail anmelden"
+    enter_email: "Bitte geben Sie Ihre E-Mail-Adresse ein"
+    submit: "Autorisieren"
+    email_sent: "Eine E-Mail wurde an %{email} mit Details zum Einloggen gesendet"
+    invalid_email: "Bitte geben Sie eine gültige E-Mail-Adresse ein"
```
Comment (phoet) [inline config/locales/de.yml:180]: we do not use "Sie" but "Du" in the rest of our communication

File: lib/omni_auth/strategies/email.rb [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+      end
+
+      def callback_phase
+        token = request.params['token']
+        fail!(:authenticity_error) if token.blank?
+
+        begin
+          decoded_token = EmailAuthToken.decode(token)
+        rescue StandardError => e
+          fail!(:authenticity_error, e)
```
Comment (phoet) [inline lib/omni_auth/strategies/email.rb:24]: This is silence the actual error and nothing will get reported to app-signal. I recommend raising the errors in the beginning, so you can see what fails and silence it later on.

File: spec/controllers/users_controller_spec.rb [hunk 10]
```diff
@@ -34,6 +34,22 @@
...
+    context 'if the user has an unfilled name' do
+      render_views
+
+      it 'clears the name input and sets it to error' do
+        allow(@controller).to receive_messages(current_user: user)
+        user.update(name: User::EMPTY_NAME)
+
+        get :edit, params: { id: user }
+
+        doc = Nokogiri::HTML(response.body)
```
Comment (phoet) [inline spec/controllers/users_controller_spec.rb:47]: there is probably some matcher for this kind of stuff, but i dont remember, so whatever.

File: app/controllers/sessions_controller.rb [hunk 11]
```diff
@@ -19,7 +21,14 @@ def create
       options = { alert: t('flash.duplicate_nick', name: e.nickname) }
     end
 
-    redirect_to request.env['omniauth.origin'].presence || root_path, options
+    redirect_path = request.env['omniauth.origin'].presence || root_path
+
+    if current_user&.missing_name?
```
Comment (phoet) [inline app/controllers/sessions_controller.rb:26]: I get the idea here, but i'm not a fan of unnecessary complexity in the signup process. it's complex as it is and users get distracted easily. i would recommend handling that in the view layer, maybe where the flash is rendered and hint to the edit page if the username has not yet been changed.

**Review by phoet — APPROVED** [review]

File: config/locales/de.yml (hunk 8 above)
Comment (josepegea) [reply config/locales/de.yml]: I speak no German (nor Polish) I'm afraid. I asked ChatGPT to do the translations hoping they would be a good starting point that could be improved by native speakers. Still, I asked again to rewrite the German texts in the colloquial form. Hoping they're not too bad!!

**Comment by josepegea** [comment]
Thanks for your review, @phoet ! 

I applied the changes you suggested.

If there's no more comments, I'll merge this tomorrow in the morning, as I'd like to have it ready for users in preparation of a meetup for next week

**Comment by josepegea** [comment]
Ok, I merged it and tried to use it. Unfortunately, it turns out that the sending of the email fails, because of missing SMTP auth. @phoet , could you please take a look at the configuration? :pray:

**Comment by josepegea** [comment]
> Ok, I merged it and tried to use it. Unfortunately, it turns out that the sending of the email fails, because of missing SMTP auth. @phoet , could you please take a look at the configuration? 🙏

I confirm that email login works if you manage to get the login link, but that's not possible as long as the app cannot send emails. I managed to make it work for myself and for one user of Madrid.rb by looking at the exception in AppSignal and forwarding the token from there. However, this is far from sustainable.

@phoet mentioned that is quite possible that the Heroku app has lost the Sendgrid configuration. I guess it used to have one, as there's another mailer that was used for usergroup notifications and the app configuration seems to point to Sendgrid. Unfortunately, I don't have access to the Heroku app, so I cannot check. @phoet is on holidays and away from a proper workplace. @salzig could you please check? :pray:

**Comment by salzig** [comment]
SENDGRID_USERNAME and SENDGRID_PASSWORD, as expected by `config/environments/production.rb` aren't configured in Heroku. Only reference I could find is SENDGRID_DOMAIN.

**Comment by salzig** [comment]
Screenshot 2024-11-02 at 20 49 55
and adding/"unlocking" the twilo SendGrid Heroku Addon ends here. <3

**Comment by josepegea** [comment]
Thanks for looking at it!  :raised_hands: 

I was just going to ask if you could enable that addon :sweat_smile:

**Comment by josepegea** [comment]
I was [looking at the docs](https://devcenter.heroku.com/articles/sendgrid#provisioning-the-add-on) and it seems that they generate a unique email address to manage the addon and require SSO from there. Maybe it was already setup?

**Comment by phoet** [comment]
I deleted and reinstalled the addon. login works, maybe emails too

**Comment by josepegea** [comment]
Hey, thanks for that. Unfortunately, I just tested and still got `SMTP-AUTH requested but missing user name` (see AppSignal, under the `ArgumentError 2081`.

Maybe it's just a matter of setting the user and password provided by the addon on the proper ENV vars: `SENDGRID_USERNAME` and `SENDGRID_PASSWORD`

**Comment by salzig** [comment]
Maybe it's cause no sender was configured. You should have received a verification mail on hello@. Can you confirm?

**Comment by josepegea** [comment]
I just checked and, yes, there was a verification email from this morning and I just clicked in the confirmation link. I'm afraid I'm in the dark regarding the configuration of the addon or the Sendgrid account as I don't have access to the Heroku dashboard

**Comment by salzig** [comment]
Screenshot 2024-11-04 at 10 29 02
maybe you where a bit late, I just resend the verification mail.

**Comment by josepegea** [comment]
Clicked it again. Still same error "550 The from address does not match a verified Sender Identity. Mail"

**Comment by salzig** [comment]
Screenshot 2024-11-04 at 10 41 42
I'm not surprised by the 550, cause send grid still shows the sender as unverified.

**Comment by josepegea** [comment]
I verified the address and registered in Sendgrid 
image
I'm really not sure if I should also "Create a new account" :thinking:

**Comment by salzig** [comment]
ah, not sure if it was the right way to create a new account. Cause maybe we can't have it verified for our send grid account now?

**Comment by josepegea** [comment]
I finalized creating the account and now no error is raised, but emails still don't arrive. Maybe now it's because of lack of SPF or similar setup for the domain, that prevent the email from being delivered. Not even in spam :thinking:

**Comment by salzig** [comment]
I don't think so, cause for Hamburg.onruby.de it works now, but the sender is verified in our account.

**Comment by josepegea** [comment]
So maybe you need to add the emails for each of the RUGs as verified senders on your account, indeed

**Comment by salzig** [comment]
Yes, I just send you a message via Discord. Maybe we can have a shortcut for "exchange" of a link there.

**Comment by salzig** [comment]
And we should introduce a "feature switch", so only accounts where the email is verified/added can use email login.

**Comment by salzig** [comment]
Madridrb sender mail is now added and verified to our send grid account. Login for HamburgOnRuby and Madridrb via Email works now.

**Comment by josepegea** [comment]
Thanks for your help, @salzig!

> And we should introduce a "feature switch", so only Whitelabel where the email is verified/added can use email login.

I'll work on that later today or tomorrow

**Comment by salzig** [comment]
Just as you suggested via Discord, maybe it's best to support a "generic" fallback via a "onruby" Global Sender first. So the feature at least works for all Labels. After that we can figure out how to support individual senders (verfied in a single account, different sendgrid account per label, etc)

**Pull request merged**